File: EtherCAT-Project/ethercat-npm/src/esiParser.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { XMLParser } from "fast-xml-parser";

export type ESIEntry = {
  filepath: string;
  metadata: Record<string, any>;
  object_dictionary: Array<Record<string, any>>;
  pdos: {
    tx: Array<Record<string, any>>;
    rx: Array<Record<string, any>>;
  };
};

const parser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: "",
});

export function loadESI(filepath: string, basePath?: string): ESIEntry {
  let file = filepath;
  if (basePath && !file.startsWith("/")) {
    file = resolve(basePath, filepath);
  }
  const xml = readFileSync(file, "utf-8");
  const json = parser.parse(xml);
  const root = json?.EtherCATInfo || json;
  const metadata = {
    vendor: root?.Vendor?.Name,
    name: root?.Descriptions?.Devices?.Device?.Type,
    revision: root?.Descriptions?.Devices?.Device?.RevisionNo,
  };
  const object_dictionary = parseObjectDictionary(root);
  const pdos = parsePdos(root);
  return { filepath: file, metadata, object_dictionary, pdos };
}

function parseObjectDictionary(root: any): Array<Record<string, any>> {
  const objects = root?.Descriptions?.Devices?.Device?.Dictionary?.Object;
  if (!objects) return [];
  const array = Array.isArray(objects) ? objects : [objects];
  return array.map((obj: any) => ({
    index: obj?.Index,
    name: obj?.Name,
    type: obj?.Type,
    subindexes: normalizeArray(obj?.SubItem).map((sub: any) => ({
      subindex: sub?.SubIndex,
      name: sub?.Name,
      bit_length: sub?.BitLen,
    })),
  }));
}

function parsePdos(root: any) {
  const pdos = { tx: [] as Array<Record<string, any>>, rx: [] as Array<Record<string, any>> };
  const profiles = normalizeArray(root?.Descriptions?.Devices?.Device?.Sm);
  // Fallback: look directly at TxPdo/RxPdo elements
  for (const direction of ["TxPdo", "RxPdo"]) {
    const entries = normalizeArray(root?.Descriptions?.Devices?.Device?.Pdo?.[direction]);
    const mapped = entries.map((pdo: any) => ({
      index: pdo?.Index,
      name: pdo?.Name,
      entries: normalizeArray(pdo?.Entry).map((entry: any) => ({
        index: entry?.Index,
        subindex: entry?.SubIndex,
        bit_length: entry?.BitLen,
        name: entry?.Name,
      })),
    }));
    (direction === "TxPdo" ? pdos.tx : pdos.rx).push(...mapped);
  }
  return pdos;
}

function normalizeArray<T>(value: T | T[] | undefined): T[] {
  if (!value) return [];
  return Array.isArray(value) ? value : [value];
}

File: EtherCAT-Project/ethercat-python/src/ethercat_mcp/esi_parser.py
"""ESI (EtherCAT Slave Information) parser helpers."""

from __future__ import annotations

from functools import lru_cache
from pathlib import Path
from typing import Any, Dict, List, Optional

from lxml import etree


class ESIParserError(RuntimeError):
    """Raised when an ESI file cannot be parsed."""


class ESIParser:
    def __init__(self, base_path: Optional[Path | str] = None) -> None:
        self.base_path = Path(base_path).expanduser() if base_path else None

    def _resolve(self, filepath: str | Path) -> Path:
        path = Path(filepath).expanduser()
        if path.is_file():
            return path
        if self.base_path:
            candidate = self.base_path / path.name
            if candidate.is_file():
                return candidate
        raise FileNotFoundError(f"ESI file not found: {filepath}")

    def load(self, filepath: str | Path) -> Dict[str, Any]:
        path = self._resolve(filepath)
        try:
            tree = etree.parse(str(path))
        except Exception as exc:
            raise ESIParserError(f"Failed to parse ESI file {path}: {exc}") from exc
        root = tree.getroot()
        metadata = {
            "vendor": root.attrib.get("Vendor"),
            "name": root.attrib.get("Name"),
            "revision": root.attrib.get("Revision"),
        }
        object_dictionary = self._parse_object_dictionary(root)
        pdos = self._parse_pdos(root)
        return {
            "filepath": str(path),
            "metadata": metadata,
            "object_dictionary": object_dictionary,
            "pdos": pdos,
        }

    def _parse_object_dictionary(self, root: etree._Element) -> List[Dict[str, Any]]:
        entries: List[Dict[str, Any]] = []
        for obj in root.findall(".//Object"):
            entries.append(
                {
                    "index": obj.attrib.get("Index"),
                    "name": obj.attrib.get("Name"),
                    "type": obj.attrib.get("Type"),
                    "subindexes": [
                        {
                            "subindex": sub.attrib.get("SubIndex"),
                            "name": sub.attrib.get("Name"),
                            "bit_length": sub.attrib.get("BitLen"),
                        }
                        for sub in obj.findall(".//SubItem")
                    ],
                }
            )
        return entries

    def _parse_pdos(self, root: etree._Element) -> Dict[str, List[Dict[str, Any]]]:
        pdos = {"tx": [], "rx": []}
        for direction, key in [("TxPdo", "tx"), ("RxPdo", "rx")]:
            entries = root.findall(f".//{direction}")
            parsed: List[Dict[str, Any]] = []
            for pdo in entries:
                mapped = [
                    {
                        "index": entry.attrib.get("Index"),
                        "subindex": entry.attrib.get("SubIndex"),
                        "bit_length": entry.attrib.get("BitLen"),
                        "name": entry.attrib.get("Name"),
                    }
                    for entry in pdo.findall(".//Entry")
                ]
                parsed.append(
                    {
                        "index": pdo.attrib.get("Index"),
                        "name": pdo.attrib.get("Name"),
                        "entries": mapped,
                    }
                )
            pdos[key] = parsed
        return pdos

    @lru_cache(maxsize=64)
    def load_cached(self, filepath: str | Path) -> Dict[str, Any]:
        return self.load(filepath)
